[FEATURE] [Pix Admin] Ajouter en masse des tags d'organisations (PIX-13456)

lego-technix opened the pull request:
## :unicorn: Problème

Jusqu'à présent on passait par un script nécessitant un accès à la production pour la création en masse de tags à des organisations.

## :robot: Proposition

Pour rendre l'exécution de cette action plus souple on met en place une nouvelle action dans Pix Admin, actuellement uniquement disponible pour les `SUPER_ADMIN`.

La protection par défaut sur la taille maximum des payloads est de 1Mb, ce qui correspond à environ un peu de plus 5000 lignes d’associations de tag par organisation.

## :rainbow: Remarques

### Classe NotFoundError

Changement de forme du constructeur de la classe `NotFoundError` pour pouvoir accueillir une propriété `meta` comme quasiment toutes les `BaseHttpError`. L'ancienne forme `new NotFoundError(message, code, title)` n'était pas/plus utilisée dans le code.

## :100: Pour tester

1. Se connecter à Pix Admin
2. Aller dans la section _Ajout de tags en masse sur des organisations_ et y soumettre un fichier CSV, par exemple : 

```csv
Organization ID,Tag name
2023,COLLEGE
2023,PUBLIC
```

3. Constater que l'opération affiche une notification de succès (en vert dans Pix Admin) et crée bien pour l'organisation `Collège House of The Dragon` les tags `PUBLIC`, ` COLLEGE`
4. Réimporter dans Pix Admin le même fichier CSV
5. Constater que l'opération se déroule sans erreur sans créer de nouveaux tags

lego-technix committed:
feat(admin): add organization tags import UI

lego-technix committed:
feat(api): add addTagsToOrganizations to organization.admin.controller

lego-technix committed:
feat(api): add new route /api/admin/organizations/import-tags-csv

lego-technix committed:
feat(api): create a TagNotFoundError

lego-technix committed:
feat(api): make it possible to pass a meta to NotFoundError

The form of the constructor with "title" was not used in the code anymore.

pix-bot-github commented:
Une fois les applications déployées, elles seront accessibles via les liens suivants :
  * [App (.fr)](https://app-pr9609.review.pix.fr)
  * [App (.org)](https://app-pr9609.review.pix.org)
  * [Orga (.fr)](https://orga-pr9609.review.pix.fr)
  * [Orga (.org)](https://orga-pr9609.review.pix.org)
  * [Certif (.fr)](https://certif-pr9609.review.pix.fr)
  * [Certif (.org)](https://certif-pr9609.review.pix.org)
  * [Junior](https://junior-pr9609.review.pix.fr)
  * [Admin](https://admin-pr9609.review.pix.fr)
  * [API](https://api-pr9609.review.pix.fr/api/)
  * [Audit Logger](https://pix-audit-logger-review-pr9609.osc-fr1.scalingo.io/api/)

Les variables d'environnement seront accessibles via les liens suivants :
  * [scalingo front](https://dashboard.scalingo.com/apps/osc-fr1/pix-front-review-pr9609/environment)
  * [scalingo api](https://dashboard.scalingo.com/apps/osc-fr1/pix-api-review-pr9609/environment)
  * [scalingo audit-logger](https://dashboard.scalingo.com/apps/osc-fr1/pix-audit-logger-review-pr9609/environment)

frinyvonnick commented on api/src/organizational-entities/domain/usecases/add-tags-to-organizations.usecase.js:
```diff
@@ -0,0 +1,79 @@
+  ],
+};
+
+export async function addTagsToOrganizations({
```
**thought**: Je trouve ça dommage de pas tirer parti du model OrganizationForAdmin qui embarque toute la logique pour mettre à jour les tags. La methode update du repository organizationForAdmin gère la persistence des tags pour ce model. Via une transaction c'est pas grave de faire plusieurs update d'affilés si je dis pas de bêtises 😄  On aurait pu rajouter une méthode findBy dans le repository pour aller chercher les models associés puis appeler les méthodes de ces models pour les mettre à jour et finalement les persister. J'ai le sentiment qu'ici on recode des mécanismes qui existent déjà 🤔 

(Hésitez pas à ignorer mon message j'ai vu de la lumière alors j'ai passé une tête 👋)

lego-technix replied:
Je regarde ça.

lego-technix replied:
Au final je n'ai pas utilisé dans cette première version du usecase `OrganizationForAdmin`. Mais des évolutions de la fonctionnalité sont prévues (suppression de tags) et là je pense que nous utiliserons `OrganizationForAdmin` en faisant une méthode dédiée pour manipuler uniquement les tags car là dans `OrganizationForAdmin` il y a uniquement une méthode `updateWithDataProtectionOfficerAndTags` qui fait trop de choses pour ce usecase.

lego-technix committed:
test(api): move an integration test file into the proper directory

This file truly hosts integration tests because thoses tests setup an HTTP
server.

mariannebost commented on api/src/organizational-entities/application/http-error-mapper-configuration.js:
```diff
@@ -23,6 +24,10 @@ const organizationalEntitiesDomainErrorMappingConfiguration = [
+  {
+    name: TagNotFoundError.name,
+    httpErrorFn: (error) => new HttpErrors.NotFoundError(error.message, error.code, error.meta),
+  },
```
Pour les autres fichiers de mapper il existe des tests unitaires qui s'assurent que le mapping se fait bien et que la bonne erreur est créée. Ca vaut le coup de l'ajouter pour ce fichier également je pense !

mariannebost commented on api/src/organizational-entities/domain/usecases/add-tags-to-organizations.usecase.js:
```diff
@@ -0,0 +1,79 @@
+import { createReadStream } from 'node:fs';
+
+import { OrganizationTag } from '../../../../lib/domain/models/OrganizationTag.js';
+import { CsvColumn } from '../../../../lib/infrastructure/serializers/csv/csv-column.js';
```
En cherchant le fichier `csv-column.js` _(car j'ignorais son existence_ 🫣) j'ai vu qu'il en existait une copie dans `src` (surement un oubli de suppression lors de sa migration) mais je pense que c'est mieux d'utiliser celui ci du coup : `api/src/shared/infrastructure/serializers/csv/csv-column.js`

mariannebost commented on api/src/organizational-entities/domain/usecases/add-tags-to-organizations.usecase.js:
```diff
@@ -0,0 +1,79 @@
+  }
+}
+
+async function retrieveTagsByName({ tagNames, tagRepository }) {
```
J'ai l'impression que c'est une méthode privée, on peut donc la renommer `_retrieveTagsByName`

mariannebost commented on api/src/organizational-entities/domain/usecases/add-tags-to-organizations.usecase.js:
```diff
@@ -0,0 +1,79 @@
+
+  const csvParser = new CsvParser(buffer, CSV_HEADER);
+  const csvData = csvParser.parse();
+  const tagNames = csvData.map(({ tagName }) => {
```
On en avait discuter avec Eric sur une ancienne PR de mettre l'étape du parsing de csv dans le controller ou dans le usecase.
Et on avait conclu que le usecase ne devrait recevoir que des données déjà bien formater pour un traitement métier uniquement et voir le parsing comme un deserialize finalement, qu'en pensez-vous ?

mariannebost commented on api/src/organizational-entities/application/organization/organization.admin.route.js:
```diff
@@ -126,4 +126,25 @@ export const organizationAdminRoutes = [
+      tags: ['api', 'admin', 'organizations', 'tags'],
+      notes: [
+        "- **Cette route est restreinte aux utilisateurs authentifiés ayant les droits d'accès**\n" +
+          '- Elle permet d’ajouter en masse des tags à des organisations.',
```
Je crois qu'il manque une partie !
`Cette route est restreinte aux utilisateurs authentifiés ayant un rôle SUPER_ADMIN permettant un accès à l'application d'administration de Pix`

mariannebost commented on admin/app/components/administration/organization-tags-import.hbs:
```diff
@@ -0,0 +1,15 @@
+    <h2 class="page-section__title">{{t "components.administration.organization-tags-import.title"}}</h2>
+  </header>
+  <p>{{t "components.administration.organization-tags-import.description"}}</p>
+  <br />
```
J'en ai vu plusieurs dans l'application et je me demande ce qui est le mieux entre utiliser une balise `<br/>` ou mettre du margin ? 🤔

lego-technix replied:
Clairement ajouter du `<br/>` pour faire de la présentation n'est pas la bonne manière de faire. Je n'ai fait que copier ce qui existait déjà, mais je suis d'accord que l'existant avec les `<br/>` n'est pas bien !

mariannebost reviewed (commented):
Testé sur Chrome avec succès ✅

lego-technix committed:
feat(api): add organizationTagCsvParser

lego-technix committed:
feat(api): add add-tags-to-organizations usecase

lego-technix committed:
refactor(api): remove now obsolete script add-tags-to-organizations.js and its tests

lego-technix committed:
refactor(admin): replace <br /> used for layout by CSS

lego-technix committed:
refactor(admin): unify fieldset+legend use

lego-technix committed:
refactor(admin): fix linting

bpetetot commented on api/src/organizational-entities/domain/usecases/add-tags-to-organizations.usecase.js:
```diff
@@ -0,0 +1,54 @@
+      throw new TagNotFoundError({ tagName });
+    }
+    tagByNames.set(tagName, tag);
+  }
```
**suggestion** simplifier la syntaxe en utilisant `for...of` (possible également à la ligne 15 sur l'organisation)

```suggestion
  for (const tagName of uniqTagNames) {
    const tag = await tagRepository.findByName({ name: tagName });

    if (tag === null) {
      throw new TagNotFoundError({ tagName });
    }
    tagByNames.set(tagName, tag);
  }
```

bpetetot commented on api/src/organizational-entities/domain/services/OrganizationTagCsvParser.js:
This should be in the `infrastructure` layer according the clean architecture.

bpetetot commented on api/tests/organizational-entities/acceptance/application/organization/organization.admin.route.test.js:
```diff
@@ -703,4 +703,50 @@ describe('Acceptance | Organizational Entities | Application | Route | Admin | O
+        ${secondOrganizationId},${thirdTag.name}
+        `;
+
+        const user = await insertUserWithRoleSuperAdmin();
```
**suggestion** ça devrait être dans le `beforeEach` et utiliser un `databaseBuilder` comme les autres données d'initialisation du test, non?

lego-technix replied:
`insertUserWithRoleSuperAdmin()` est en fait un raccourci appelant le `databaseBuilder`.  Par contre ta remarque m'a montré qu'il manquait un `databaseBuilder.commit()`.

Et plus généralement tout ce fichier de test n'était pas rationnel dans la création de l'utilisateur _admin_ et il manquait parfois à d'autres endroits des `databaseBuilder.commit()`. J'ai donc ajouté à cet PR un dernier commit pour factoriser proprement la création de l'utilisateur _admin_ qui sert à la grande majorité des tests de ce fichier.

bpetetot commented on admin/translations/en.json:
```diff
@@ -65,6 +65,17 @@
         "upload-button": "Import JSON file"
       },
+      "organization-tags-import": {
+        "title": "Mass addition of tags on organisations",
```
```suggestion
"title": "Bulk addition of tags on organisations",
```

mariannebost reviewed (approved):
Très bonne idée le `OrganizationTagCsvParser` !

lego-technix committed:
test(api): fix some missing databaseBuilder.commit()

through factoring out admin creation in top beforeEach
